hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/SignalUtil.java

```diff
@@ -0,0 +1,152 @@
+      BindingUtils.loadClassSafely("sun.misc.Signal");
+  static final Class<?> jdkSignalHandlerClazz =
+      BindingUtils.loadClassSafely("sun.misc.SignalHandler");
+
+  static DynConstructors.Ctor<?> jdkSignalCtor =
+      new DynConstructors.Builder()
+          .impl(jdkSignalClazz, String.class)
+          .build();
+
+  static DynMethods.StaticMethod jdkSignalHandleStaticMethod =
+      new DynMethods.Builder("handle")
+          .impl(jdkSignalClazz, jdkSignalClazz, jdkSignalHandlerClazz)
+          .buildStatic();
+
+  static DynMethods.StaticMethod jdkSignalRaiseStaticMethod =
+      new DynMethods.Builder("raise")
+          .impl(jdkSignalClazz, jdkSignalClazz)
+          .buildStatic();
+
+  static DynMethods.UnboundMethod jdkSignalHandlerHandleMethod =
+      new DynMethods.Builder("handle")
+          .impl(jdkSignalHandlerClazz, jdkSignalClazz)
+          .build();
+
+  @InterfaceAudience.Private
+  public static class Signal {
+    private final static DynMethods.UnboundMethod getNumberMethod =
+        new DynMethods.Builder("getNumber").impl(jdkSignalClazz).build();
+
+    private final static DynMethods.UnboundMethod getNameMethod =
+        new DynMethods.Builder("getName").impl(jdkSignalClazz).build();
+
+    private final Object delegate;
+
+    public Signal(String name) {
+      this.delegate = jdkSignalCtor.newInstance(name);
+    }
+
+    public Signal(Object delegate) {
+      Preconditions.checkArgument(jdkSignalClazz.isInstance(delegate),
+          String.format("Expected class is '%s', but actual class is '%s'",
+              jdkSignalClazz.getName(), delegate.getClass().getName()));
+      this.delegate = delegate;
+    }
+
+    public int getNumber() {
+      return getNumberMethod.bind(delegate).invoke();
+    }
+
+    public String getName() {
+      return getNameMethod.bind(delegate).invoke();
+    }
+
+    public boolean equals(Object obj) {
+      if (this == obj) {
+        return true;
+      }
+      if (obj instanceof Signal) {
+        return delegate.equals(((Signal)obj).delegate);
+      }
+      return false;
+    }
+
+    public int hashCode() {
+      return delegate.hashCode();
+    }
+
+    public String toString() {
+      return delegate.toString();
+    }
+  }
+
+  @InterfaceAudience.Private
+  public interface Handler {
+    void handle(Signal sig);
+  }
+
+  static class JdkSignalHandlerImpl implements Handler {
+
+    private final Object delegate;
+
+    JdkSignalHandlerImpl(Handler handler) {
+      this.delegate = Proxy.newProxyInstance(
+          getClass().getClassLoader(),
+          new Class<?>[] { jdkSignalHandlerClazz },
+          (proxyObj, method, args) -> {
+            if ("handle".equals(method.getName()) && args.length == 1 && jdkSignalClazz.isInstance(args[0])) {
+              handler.handle(new Signal(args[0]));
+              return null;
+            } else {
+              return InvocationHandler.invokeDefault(proxyObj, method, args);
```

Comment:
`InvocationHandler.invokeDefault` was only introduced after Java 16. Will this affect compilation with `-release 8`?